plugin/core/sessions.py

```diff
@@ -621,24 +658,40 @@ def m_textDocument_publishDiagnostics(self, params: Any) -> None:
 
     def m_client_registerCapability(self, params: Any, request_id: Any) -> None:
         """handles the client/registerCapability request"""
-        registrations = params["registrations"]
-        for registration in registrations:
-            method = registration["method"]
-            capability_path, registration_path = method_to_capability(method)
-            debug("{}: registering capability:".format(self.config.name), capability_path)
-            self.capabilities.set(capability_path, registration.get("registerOptions", {}))
-            self.capabilities.set(registration_path, registration["id"])
-        self.send_response(Response(request_id, None))
+
+        def run() -> None:
+            registrations = params["registrations"]
+            for registration in registrations:
+                method = registration["method"]
+                capability_path, registration_path = method_to_capability(method)
+                debug("{}: registering capability:".format(self.config.name), capability_path)
+                self.capabilities.set(capability_path, registration.get("registerOptions", {}))
+                self.capabilities.set(registration_path, registration["id"])
+                toplevel_key = capability_path.split('.')[0]
+                if toplevel_key.endswith('Provider'):
+                    for sv in self.session_views_async():
+                        sv.register_capability(toplevel_key)
+            self.send_response(Response(request_id, None))
+
+        sublime.set_timeout_async(run)
```

Comment:
Does running those async can potentially cause a race issue where a server would expect those to be registered already when it sends the next (one that relies on registration) message?

Reply:
The server can only be sure we're done with our bookkeeping once we have replied to the request, which we do after we did all the bookkeeping. So everything checks out I think.